### PR TITLE
Feature/#176 create allergens section

### PR DESCRIPTION
**Pull request by MChimali**
### **`/Backend` modifications:**
1. In `/dals/restaurant/restaurant.model`:
      - Created `AllergenEnum` type with all allergens types.
      - In Item created optional `alllergenCollection` property.
      
2. In `/dals/restaurant/mock-data`:
      - Added to restaurant data `allergenCollection` property into the menus of the restaurants.

3. In /pods/restaurant:
      - In `restaurant.api-model` created `AllergenEnum` type and `allergenCollection` property in Item.
      - In `restaurant.mappers` added property `allergenCollection ` to `mapFromItemApiToItemModel` and `mapFromItemModelToItemModelApi`
      
### **`/menu-public-portal` modifications:**
1. In /public created allergen-icons folder with:
      - webp folder with low quality .webp files to display allergens in menu.
      - png folder with medium quality .png files to display allergens symbols in legend.
      - Created `allergen-icons.list` with a list pery type of image with objects containing name/localURL.
     
2. In `pods/restaurant/api/restaurant.api.model`:
      - Created `AllergenEnum` type with all allergens types.
      - In Item created optional `alllergenCollection` property.
      
3. In `pods/restaurant/restaurant.vm`:
      - Created `AllergenInfo` interface with name and localURL to image.
      - In Item added property `alllergenCollection?: AllergenInfo[]`.
      
4. In pods/restaurant/restaurant.mapper:
      - Added to `mapFromItemsApiToItemsVm` property `allergenIconsUrl` to be from `allergenCollection`.
      - Added `mapAllergenCollectionToAllergenIconsUrl` and `AllergenToURLobject` to map `allergenCollection` into `allergenIconsUrl`.

5. Added tests to `pods/restaurant/restaurant.mapper.spec`.

6. In `pods/restaurant/restaurant.component`: 
      - In `DishesComponent` added `AllergensComponent` to display allergens icons in each dish, if available.
      - If `AllergenIconsUrl `is null > warning message displayed.
      - If `AllergenIconsUrl` is `[]` empty array > food with no allergens.
      - Created `AllergensLegend `component to display wrapped in `Accordion `component to display in legend `.png` icons.

### CODE REVIEW

File: menu-public-portal/package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "@mui/icons-material": "^5.3.1",
     "@mui/material": "^5.4.0",
     "next": "12.1.0",
+    "next-images": "^1.8.4",
```
Comment (manudous) [inline menu-public-portal/package.json:27]: do you need it?

File: menu-public-portal/src/pods/restaurant/api/restaurant.api.ts [hunk 2]
```diff
@@ -6,6 +6,9 @@ export const getRestaurantByName = async (
 ): Promise<RestaurantApi> => {
   const restaurant = await fetch(
     `${envConstants.BASE_API_URL}/restaurant/${restaurantName}`
-  ).then((response) => response.json());
+  ).then((response) => {
+    return response.json();
+  });
+  console.log('***1', restaurant.menu[0], restaurant.allergenCollection);
```
Comment (manudous) [inline menu-public-portal/src/pods/restaurant/api/restaurant.api.ts:12]: remove console.log

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx [hunk 3]
```diff
@@ -156,6 +218,16 @@ export const RestaurantComponent: React.FC<Props> = (props) => {
...
+      <div className={classes.accordion}>
+        <Accordion>
+          <AllergenSummaryStyled>
+            <Typography>Lista de Alérgenos</Typography>
+          </AllergenSummaryStyled>
+          <AccordionDetails>
+            <AllergensLengend />
+          </AccordionDetails>
+        </Accordion>
+      </div>
```
Comment (manudous) [inline menu-public-portal/src/pods/restaurant/restaurant.component.tsx:230]: move this after first accordion

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx [hunk 4]
```diff
@@ -159,3 +231,13 @@ export const RestaurantComponent: React.FC<Props> = (props) => {
...
+var myMethod = function () {
+  console.log(this);
+};
+
+var myObject = {
+  myMethod: myMethod,
+};
+
+myObject.myMethod(); // this === myObject
+myMethod(); // this === window
```
Comment (manudous) [inline menu-public-portal/src/pods/restaurant/restaurant.component.tsx:243]: do you need this?

File: menu-public-portal/src/pods/restaurant/restaurant.mapper.ts [hunk 5]
```diff
@@ -47,8 +51,21 @@ const mapFromItemsApiToItemsVm = (
...
+    ? mapAllergenCollectionToAllergenIconsUrl(item.allergenCollection)
+    : null,
 });
 
+const mapAllergenCollectionToAllergenIconsUrl = (
+  allergenCollection: restaurantApi.AllergenEnum[]
+): AllergenInfo[] =>
+  allergenCollection.length > 0
+    ? allergenCollection.map(AllergenToURLobject)
+    : [];
```
Comment (manudous) [inline menu-public-portal/src/pods/restaurant/restaurant.mapper.ts:64]: check in menu-public-portal>src>common>mappers>collection.mappers can do this for you

**Review by MChimali — COMMENTED** [review]
Please note in menu CategoryEntry "Hamburguesas a la Brasa". I have divided "Filetes de Hamburguesas in two item groups, according to the allergens they share. 
- "Filete de Hamburguesa mixta / Filete de Hamburguesa hereford" 
- "Filete de Hamburguesa de pollo / Filete de Hamburguesa picante".

It can also be divided into 4 different CategoryEntry items.

File: justo.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+console.log(true == 1);
```
Comment (brauliodiez) [inline justo.js:1]: What is this? justo.js?

File: backend/src/dals/restaurant/restaurant.model.ts [hunk 7]
```diff
@@ -16,6 +32,7 @@ export interface Item {
   price?: number;
   priceByRation?: SubItemPrice;
   unit?: string;
+  allergenCollection?: AllergenEnum[];
```
Comment (brauliodiez) [inline backend/src/dals/restaurant/restaurant.model.ts:35]: @manudous and @Nasdan I think Allergen Collection should place by Item not by ration type, any ration will contain the same allergens, please review this, I have added the field into moon Modeler.

Just to keep the name consistent on moonmodeller should we call it _allergens_ : string[] ?

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx [hunk 8]
```diff
@@ -5,14 +5,66 @@ import AccordionDetails from '@mui/material/AccordionDetails';
...
+              <Image src={item.iconUrl} height={35} width={35} />
+            </div>
+          ))}
+        </div>
+      ) : (
+        <Typography
+          variant="subtitle1"
+          sx={{ color: 'green', marginTop: '0.25rem' }}
+        >
+          Este plato no contiene ningún alérgeno
```
Comment (brauliodiez) [inline menu-public-portal/src/pods/restaurant/restaurant.component.tsx:62]: I think we should just keep an empty line here, it's too dangeours indicating that the dish it self does not contain any "alergeno" (it's more a legal issue)

File: backend/src/dals/mock-data.ts [hunk 9]
```diff
@@ -235,82 +261,97 @@ export const db: DB = {
...
                 'Tomate, Carne, Salsa Barbacoa, Mozzarella y Orégano',
               price: 7,
+              allergenCollection: ['gluten', 'lactosa'],
             },
             {
               name: 'Calzone Manolín',
               description:
                 'Mozzarella, Tomate, Champiñones, Bacón, Queso Roquefort, Chorizo y Huevo',
               price: 7,
+              allergenCollection: ['gluten', 'lactosa', 'huevos'],
```
Comment (brauliodiez) [inline backend/src/dals/mock-data.ts:354]: If this is the data that is stored in the database and goes to production we should only display allergen on papulinos, for the rest of restaurants we don't have such information, we could have legal issues and people could get sick

File: backend/src/dals/restaurant/restaurant.model.ts (hunk 7 above)
Comment (brauliodiez) [reply backend/src/dals/restaurant/restaurant.model.ts:35]: Oops sorry it was wrongb from my side coded was shrinked, expanded an d it's on the right place :)

File: backend/src/dals/restaurant/restaurant.model.ts [hunk 10]
```diff
@@ -1,6 +1,22 @@
 import { ObjectId } from 'mongodb';
 import { RestaurantTheme } from 'common-app/models';
 
+export type AllergenEnum =
```
Comment (Nasdan) [inline backend/src/dals/restaurant/restaurant.model.ts:4]: I'd move this type to `./src/common-app/models`  as we did with `RestaurantTheme`. We can reuse this type and we don't have to declare again in `restaurant.api-model`

File: menu-public-portal/public/allergen-icons/allergen-icons.list.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
+import { AllergenEnum } from 'pods/restaurant/api';
```
Comment (Nasdan) [inline menu-public-portal/public/allergen-icons/allergen-icons.list.ts:1]: Why is this file on public folder?

**Comment by Nasdan** [comment]
Could you rename images files to kebab case, please? We could have some issues with uppercase and lowercase and git.

For example: `frutos-secos-con-cascara.png`

File: menu-public-portal/src/pods/restaurant/api/restaurant.api.ts [hunk 12]
```diff
@@ -6,6 +6,8 @@ export const getRestaurantByName = async (
 ): Promise<RestaurantApi> => {
   const restaurant = await fetch(
     `${envConstants.BASE_API_URL}/restaurant/${restaurantName}`
-  ).then((response) => response.json());
+  ).then((response) => {
```
Comment (Nasdan) [inline menu-public-portal/src/pods/restaurant/api/restaurant.api.ts:9]: Why this change?

File: menu-public-portal/src/pods/restaurant/restaurant.vm.ts [hunk 13]
```diff
@@ -1,4 +1,19 @@
 import { ThemeName } from 'core/theme';
+export type AllergenEnum =
```
Comment (Nasdan) [inline menu-public-portal/src/pods/restaurant/restaurant.vm.ts:2]: Same approach as back. Move to `menu-public-portal/src/common-app/models`

File: menu-public-portal/src/pods/restaurant/restaurant.mapper.ts [hunk 14]
```diff
@@ -47,8 +51,14 @@ const mapFromItemsApiToItemsVm = (
     ? mapFromPriceByRationApiToPriceByRationVm(item.priceByRation)
     : null,
   unit: item.unit ? item.unit : null,
+  allergenIconsUrl: item.allergenCollection
+    ? mapToCollection(item.allergenCollection, AllergenToURLobject)
+    : null,
 });
 
+export const AllergenToURLobject = (allergen): AllergenInfo =>
```
Comment (Nasdan) [inline menu-public-portal/src/pods/restaurant/restaurant.mapper.ts:59]: use cameCase

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx [hunk 15]
```diff
@@ -5,14 +5,59 @@ import AccordionDetails from '@mui/material/AccordionDetails';
...
+}
+
+const AllergensLengend: React.FC = () => {
+  return (
+    <>
+      <div className={classes.allergenLegendBox}>
+        {urlIconsPng.map((item) => (
+          <div className={classes.legendItem}>
+            <Image src={item.iconUrl} height={120} width={120} />
+            <Typography style={{ fontSize: '0.9rem' }}>{item.name}</Typography>
```
Comment (Nasdan) [inline menu-public-portal/src/pods/restaurant/restaurant.component.tsx:36]: please move styles to styles file

File: backend/src/dals/restaurant/restaurant.model.ts (hunk 7 above)
Comment (Nasdan) [inline backend/src/dals/restaurant/restaurant.model.ts:35]: Rename `allergenCollection` to `allergens` as moon modeler

File: menu-public-portal/public/allergen-icons/allergen-icons.list.ts (hunk 11 above)
Comment (MChimali) [reply menu-public-portal/public/allergen-icons/allergen-icons.list.ts:1]: I don't like it either. Where would you rather place it?

File: menu-public-portal/src/common/components/accordion-styled.tsx [hunk 16]
```diff
@@ -14,3 +15,14 @@ export const AccordionSummaryStyled = styled((props: AccordionSummaryProps) => (
...
 }));
+
+export const AllergenSummaryStyled = styled((props: AccordionSummaryProps) => (
+  <MuiAccordionSummary
+    {...props}
+    expandIcon={<ExpandMoreIcon sx={{ color: 'text.secondary' }} />}
+  />
+))(({ theme }) => ({
+  color: theme.palette.text.secondary,
+  backgroundColor: varsRestaurantType.brewery.secondaryColor,
```
Comment (Nasdan) [inline menu-public-portal/src/common/components/accordion-styled.tsx:26]: use here, `backgroundColor: theme.palette.secondary.main,`

File: menu-public-portal/src/common/components/accordion-styled.tsx [hunk 17]
```diff
@@ -3,6 +3,7 @@ import { styled } from '@mui/material/styles';
 import MuiAccordionSummary from '@mui/material/AccordionSummary';
 import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
 import { AccordionSummaryProps } from '@mui/material/AccordionSummary';
+import { breweryTheme, varsRestaurantType } from 'core/theme';
```
Comment (Nasdan) [inline menu-public-portal/src/common/components/accordion-styled.tsx:6]: Remove this import

File: menu-public-portal/public/allergen-icons/allergen-icons.list.ts [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+
+export interface AllergenInfo {
+  name: AllergenEnum;
+  iconUrl: string;
+}
+
+const baseUrlPng = '/allergen-icons/png';
+const baseUrlWebp = '/allergen-icons/webp';
+
+export const urlIconsPng: AllergenInfo[] = [
```
Comment (Nasdan) [inline menu-public-portal/public/allergen-icons/allergen-icons.list.ts:11]: Why we have 2 extensions of icons `png` and `webp`?  Why not only `webp`?

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx [hunk 19]
```diff
@@ -63,6 +108,16 @@ export const DishesComponent: React.FC<PropsItemsComponent> = (props) => {
             {item.priceByRation ? (
               <RationComponent ration={item.priceByRation} />
             ) : null}
+            {item.allergenIconsUrl ? (
+              <AllergensComponent urlAllergenList={item.allergenIconsUrl} />
+            ) : (
+              <Typography
```
Comment (Nasdan) [inline menu-public-portal/src/pods/restaurant/restaurant.component.tsx:114]: Remove this typography, if `item.allergenIconsUrl` is empty it should render nothing

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx [hunk 20]
```diff
@@ -148,6 +203,16 @@ export const RestaurantComponent: React.FC<Props> = (props) => {
           </Accordion>
         ))}
       </div>
+      <div className={classes.accordion}>
```
Comment (Nasdan) [inline menu-public-portal/src/pods/restaurant/restaurant.component.tsx:206]:
```
{ item.allergenIconsUrl.length > 0 && (
    <div className={classes.accordion}>
        <Accordion>
          <AllergenSummaryStyled>
            <Typography>Lista de Alérgenos</Typography>
          </AllergenSummaryStyled>
          <AccordionDetails>
            <AllergensLengend />
          </AccordionDetails>
        </Accordion>
      </div>
)}
```

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx (hunk 19 above)
Comment (Nasdan) [reply menu-public-portal/src/pods/restaurant/restaurant.component.tsx:114]:
```
{item.allergenIconsUrl.length > 0 && (
    <AllergensComponent urlAllergenList={item.allergenIconsUrl} />
)}
```

File: menu-public-portal/public/allergen-icons/allergen-icons.list.ts (hunk 11 above)
Comment (Nasdan) [reply menu-public-portal/public/allergen-icons/allergen-icons.list.ts:1]: As we commented, we could create `./src/pods/restaurant/restaurant.constants.ts`

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx (hunk 20 above)
Comment (MChimali) [reply menu-public-portal/src/pods/restaurant/restaurant.component.tsx:206]: This is not possible because `item.allergenIconsUrl.length` here is out of the scope of item. This is being rendered on the scope of restaurant and because of that, it will not have access to the different menus (this is where allergens are) within the restaurant.

@brauliodiez and @manudous discussed the possibility of implementing an Allergens flag in the interface of RestaurantApi. This way we render the legend of allergens only if the flag is true.

File: menu-public-portal/src/pods/restaurant/restaurant.component.tsx (hunk 19 above)
Comment (MChimali) [reply menu-public-portal/src/pods/restaurant/restaurant.component.tsx:114]: At the moment optional property allergenIconsUrl is subject to property allergens, which is also optional. Item.allergenIconsUrl.length will throw an error if allergen is null.

There is the option of mapping every single menu to an empty array allergenIconsUrl if no allergens collection is provided, but I do not like this option because I think it can be useful for future to distinct between:
-  null property allergens: restaurant did not provide allergens on menu.
- empty array allergens: menu do not contain any allergens.

File: backend/src/dals/mock-data.ts (hunk 9 above)
Comment (MChimali) [reply backend/src/dals/mock-data.ts:354]: I do not really understand this comment. The info displayed in mock-data is solely for dev purposes, right? When deploying the app the info is taken from `console-runners` which is verified information, right?
